### ISSUE TITLE
[Suggestion] Spider Features: Slowness, Webbing, and Leaping

### ISSUE THREAD

**Issue by SandwichHorror** [Insane96/EnhancedAI#6]
**What is this?**
Several enhanced AI/difficulty suggestions for spiders, one of the most forgettable Minecraft mobs.

- [Implemented in Mobs Properties Randomness] Spiders inflict Slowness on hit
This could be done with datapacks, but is possibly a prerequisite for other features:

- [Will be implemented in Mobs Properties Randomness] Spiders inflict stacking Slowness on hit
Inspiration from Rough Mobs Revamped, which never updated past 1.15. Each time a spider hits an entity, the entity's  Slowness level increases and the duration is refreshed, up to Slowness IV.

- [Don't really like the idea of placing a block at the target hit] Spiders place a web block at target's location on hit.
This could be configurable to be chance-based (0.0-1.0 per hit) or related to the above suggestions:
   Variant: Spiders place a web block after hitting an entity [X] consecutive times. Allows greater predictability of the feature than pure random chance, and encourages quick action and evasion rather than "tanking" spider attacks. 
   
- [x] Spiders can launch webs at target.
As an alternative or supplement to the above "melee webbing", this would allow spiders to pose a greater threat from a distance. When a spider is a certain distance away from its target, it fires a web projectile before moving in closer. This would require the creation of a new projectile entity and web projectile asset, which might be beyond the desired scope of Enhanced AI.
   
- [ ] Improved spider leaping
Unlike the above features, this suggestion has not been present in any mod that I know of. Improved leaping would grant spiders immunity to fall damage. It would also increase the distance covered when spiders "leap". More advanced tasks might allow spiders to "evaluate" gaps and clear gaps by leaping. Similar behavior can be seen in some datapacks such as Jumping Mobs (https://www.planetminecraft.com/data-pack/jumping-mobs-java-datapack-original-by-pheonixsc/) or mods such as Mobs Attempt Parkour (https://www.curseforge.com/minecraft/mc-mods/mobs-attempt-parkour - which is only for Fabric). 
If the mob leaping intrigues you, it could be a general feature applied to all mobs, though preferably with a blacklist or whitelist.

**Why?**
The spider is perhaps the least memorable and least interesting "classic" Minecraft monster. 
Despite being the first designed hostile mob, it has received the least attention and fewest feature updates compared to Creepers, Skeletons, or Zombies. 
The spider's main threat was its notable speed, which became irrelevant as soon as players became able to sprint in Minecraft Beta. Over the years, the spider received limited X-Ray and the ability to crudely ascend walls - but with poor AI, no fall damage immunity, and a single block or fence able to stop climbing, spiders are still immensely underwhelming.

In the modded ecosystem, mods increasing the difficulty and abilities of creepers, skeletons, and zombies are commonplace. For 1.16.5 alone, Consecration, Creeper AI Updated, Stalking Creepers, Nasty Mobs, Improved Mobs, Special AI, and others allow for significant customization of these mobs to gain new features, while spiders remain relatively forgotten. 

The greatest enhancement to spiders available for 1.16.5 is the Spiders 2.0 mod (https://www.curseforge.com/minecraft/mc-mods/spiders-2-0), but that mod has not received updates in some time and adds features only relating to the spiders' climbing and pathfinding abilities. 

These webbing and leaping features would grant spiders greater difficulty and feature parity with other mobs in Enhanced AI, and continue the legacy of Epic Siege Mod and Rough Mobs, which granted slowness and webbing features to the spider in their original incarnations.

**Comment by Insane96** [comment]
I don't really like having both slowness and webbing on attack.
The things I would like to implement is launching a web projectile that slows the player on hit and stacks (maybe even webbing them), so spiders can catch up with the Player + the slowing on hit.

> This would require the creation of a new projectile entity and web projectile asset, which might be beyond the desired scope of Enhanced AI.
That's actually what I wanted to do

> Improved spider leaping
This is a nice suggestion, although it might be a good feature for every mob, I think every mob should have unique abilities, so apply it only to spiders.

**Comment by SandwichHorror** [comment]
Yeah, the slowness and webbing on attack quite strong, but if you didn't want to make a projectile, I think that is way more balanced than other mods/datapacks that make the spider place a cobweb on EVERY hit - that Rough Mobs method would require getting hit 3-4 consecutive times before being webbed. 

However, the slowness on hit plus a web projectile takes more effort on the modder's part, but is way more interesting. I'm very glad that's in line with what you had planned.

As for the leaping, I think it could be a separate "module", like how the targeting is one "module", but you'd have it so that by default only spiders have that ability.

Or even if it's listed under the spider section, something like

```
# Leaping
# Determines which entities can pathfind over several-block gaps via leaping (default: minecraft:spider)
leaping: ["minecraft:spider"]
```

The fall damage immunity could even be a separate ability, so users who want their zombies and creepers and modded mobs to leap over blocks could still have them potentially die from falling. 

IMHO, the absolute best method of a mob-enhancing mod like Enhanced AI is to be shipped with default options for your "vision" (such as every mob having unique abilities, stronger abilities being rarer percentage-wise, most special abilities are chance-based) while still allowing the user to configure the mod to create their desired experience (such as certain features being guaranteed or disabled, multiple different mobs sharing a certain formerly "unique" ability, etc.)

**Comment by SandwichHorror** [comment]
Yeah, that probably isn't the best wording. I don't know exactly how it works, but I just meant "leap over a gap when a mob wants to move or reach something on the other side", which is my understanding of what that linked datapack and mod provided.

**Comment by Insane96** [comment]
Spiders can launch webs at target (mixed with Spiders inflict stacking Slowness on hit) will be available in Alpha 1.2.0

https://user-images.githubusercontent.com/7185104/134798853-4e6da92c-42b2-4d85-a85d-6b400aa20938.mp4

**Comment by Insane96** [comment]
> no fall damage immunity

That's something I'll definitely add. #36

**Comment by Insane96** [comment]
Closing this in favor of #208
